### ISSUE TITLE
ENH: Allow `ndarray.shape` to be a named tuple

### ISSUE THREAD

**Issue by jorenham** [numpy/numpy#27086]
### Proposed new feature or change:

Curently, it's allowed to pass some `typing.NamedTuple` instance to e.g. `numpy.empty` or `numpy.zeros`, but somewhere along the way, it gets converted to a vanilla `tuple`:

```pycon
>>> from typing import NamedTuple
>>> import numpy as np
>>> class Shape2D(NamedTuple):
...     x: int
...     y: int
... 
>>> x = np.zeros(Shape2D(x=22, y=7))
>>> type(x.shape)
<class 'tuple'>
```

I think it would be cool if you could have `ndarray.shape` be a named tuple. 

From a typing perspective, it also makes sense to have the `ShapeType` parameter of `ndarray[ShapeType: tuple[int, ...], DType]` match `ndarray.shape: ShapeType`, in case it isn't exactly a `tuple` but a subtype like `NamedTuple`.

**Comment by seberg** [comment]
If shape would have names, that would be much like a labeled array like `xarray`.  I wouldn't mind NumPy supporting labeled axes!
But the important point about labeled axes is that you have to take the names into account when you do things like broadcasting.
(Dunno if named tuples make sense, because you need a type for each axes combination, but maybe you can stuff it somewhere if a user provides it.)

**Comment by mattip** [comment]
Isn't this [structured arrays](https://numpy.org/devdocs/user/basics.rec.html#structured-arrays)? They are less performant than regular arrays.

**Comment by rkern** [comment]
No, this would be different, about labeling axes rather than fields within each item.

**Comment by jorenham** [comment]
@seberg Perhaps it'd be easiest to just convert the named `shape` into a vanilla `tuple` in those cases. 
Or alternatively; only allow broadcasting `x` and `y` if `isinstance(x.shape, type(y.shape))` or vice versa 🤔

**Comment by seberg** [comment]
> Or alternatively; only allow broadcasting `x` and `y` if `isinstance(x.shape, type(y.shape))` or vice versa

TBH, at that point the whole idea is so limiting compared to labeled axes that I don't think it's worthwhile.  You should be able to define things like `arr[0]` to remove one name at least.  Ideally, you are able to define result type names for simple broadcasting too.

If you solve both of those problems `arr[0].shape` and at least `(arr[0] + arr).shape` or so, then I see a start of something nice here.  I am not sure if that can work well with named tuples or not (maybe, there may be worse things than dynamically creating named tuples).

**Comment by jorenham** [comment]
It's too bad that dynamically created namedtuples (or any types that are created at for that matter) can't be statically typed at the moment.

But from a typing perspective, that shouldn't matter too much:
Because currently there's also no way of typing the shape of e.g. `arr[2]` given `arr: ndarray[tuple[L[480], L[720], L[3]], dtype[uint8]]`. 
It isn't impossible, but positive integer indices, this would require an additional 64 overloads (because passing the `_ShapeType: tuple[int, ...]` type-var to `typing.Unpack` is invalid, but I digress).
